### PR TITLE
39 sliding window workflow

### LINKED ISSUE #39: MVP example for model inference and saving predicted results.
Sliding window inference
Writing segmentation as a nifti image (Nifti image writer)
Writing classification as a json file (json writer)

### PR DESCRIPTION
**Pull request by wyli**
Fixes #39 #6 #5 .

### Description
combining PR #52 #53 #57  implementing a sliding window inference workflow in `examples/unet_inference_3d.py`

TODOs
~- unit test for event handler: `monai/application/handlers/segmentation_saver.py`~
~- integration test for the workflow~
(advanced testing tasks should be part of #17)
- [ ] discussion of the niftidataset changes for metadata `monai/data/readers/niftireader.py`

- [x] implement a model checkpoint loading event handler?

### Status
**Ready**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### CODE REVIEW

File: monai/data/readers/niftireader.py [hunk 1]
```diff
@@ -97,4 +103,14 @@ def __getitem__(self, index):
             random.seed(seed)  # ensure randomized transforms roll the same values for segmentations as images
             seg = self.seg_transform(seg)
 
-        return img, seg
+        if self.image_only or meta_data is None:
+            return img, seg
+
+        compatible_meta = {}
```
Comment (wyli) [inline monai/data/readers/niftireader.py]: to get the corresponding input file name when saving a segmentation output, I need to change the current `NiftiDataset ` class to return metadata. What's implemented here is a quick workaround. 

this is related to the input pipeline design, any comments/suggestions? @ericspod @atbenmurray @Nic-Ma  

dtypes of some of the metadata items are not supported by the default collate function of pytorch dataloader, I have to put a filter here to return only the compatbile metadata.

File: examples/unet_inference_3d.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+)
+net.to(device)
+
+
+def _sliding_window_processor(_engine, batch):
+    net.eval()
+    img, seg, meta_data = batch
+    with torch.no_grad():
+        seg_probs = sliding_window_inference(img, roi_size, sw_batch_size, net, device)
+        return predict_segmentation(seg_probs)
```
Comment (wyli) [inline examples/unet_inference_3d.py]: the segmentation prob to crisp mask conversion is implemented here in the workflow, after stitching the patches. so I think the network's output quantisation https://github.com/Project-MONAI/MONAI/blob/294f6e13a70fe813aa35284e25b8019a3653cc97/monai/networks/nets/unet.py#L101 should be moved into an independent post-processing module, what do you think? @ericspod @atbenmurray  (this is related to issue #44)

File: monai/utils/sliding_window_inference.py [hunk 3]
```diff
@@ -9,20 +9,22 @@
...
 
 def sliding_window_inference(inputs, roi_size, sw_batch_size, predictor, device):
     """Use SlidingWindow method to execute inference.
 
     Args:
+        inputs (numpy array): input image to be processed (assuming NCHW[D])
         roi_size (list, tuple): the window size to execute SlidingWindow inference.
         sw_batch_size (int): the batch size to run window slices.
-        predictor: a portion from numpy.lib.arraypad.pad is copied below.
+        predictor: a moani.networks.nets module
```
Comment (Nic-Ma) [inline monai/utils/sliding_window_inference.py]: monai, not "moani".

File: tests/test_sliding_window_inference.py [hunk 4]
```diff
@@ -10,27 +10,36 @@
...
-        device = torch.device("cuda:0")
+    @parameterized.expand([TEST_CASE_1, TEST_CASE_2, TEST_CASE_3, TEST_CASE_4])
+    def test_sliding_window_default(self, image_shape, roi_shape, sw_batch_size):
+        inputs = np.ones(image_shape)
+        device = torch.device("cpu:0")
 
         def compute(data):
-            data = torch.from_numpy(data)
-            return data.to(device) + 1
+            # data = torch.from_numpy(data)
```
Comment (Nic-Ma) [inline tests/test_sliding_window_inference.py]: Remove this line directly.

File: monai/data/writers/niftiwriter.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        revert_canonical (bool): whether to revert canonical.
+        dtype (np.dtype, optional): convert the loaded image to this data type.
+
+    """
+    assert isinstance(data, np.ndarray), 'input data must be numpy array.'
+    if affine is None:
+        affine = np.eye(4)
+
+    if revert_canonical:
+        codes = nib.orientations.axcodes2ornt(nib.orientations.aff2axcodes(np.linalg.inv(affine)))
```
Comment (wyli) [inline monai/data/writers/niftiwriter.py]: a follow-up of https://github.com/Project-MONAI/MONAI/pull/53#discussion_r375914183 I don't fully understand this reverting, could we chat with the original author? @Nic-Ma

File: examples/unet_inference_3d.py (hunk 2 above)
Comment (ericspod) [reply examples/unet_inference_3d.py]: I would say that people would expect a segmentation network to produce segmentations itself without processing, rather than just produce the logits. The network should be all-inclusive of its functionality, certainly if we package it with torchscript  users would expect it to have all the functionality baked in to accomplish the task.

Comment (wyli) [reply examples/unet_inference_3d.py]: just noticed that this unet implementation doesn't have a final softmax/sigmoid layer for segmentation probs, which means it is a generic arch. could be used for image regression such as MR-CT synthesis? let's have a meeting early next week to discuss it #44

File: monai/utils/arrayutils.py [hunk 6]
```diff
@@ -219,10 +219,61 @@ def iter_patch_slices(dims, patch_size, start_pos=()):
         yield tuple(slice(s, s + p) for s, p in zip(position[::-1], patch_size))
 
 
+def iter_dense_patch_slices(image_size, patch_size, scan_interval):
```
Comment (ericspod) [inline monai/utils/arrayutils.py]: This function isn't a generator, it returns a list. The name and docstring should be changed to match, though I would suggest we modify iter_patch_slices to fill this role instead. I was planning to look into that shortly.

File: monai/data/writers/niftiwriter.py (hunk 5 above)
Comment (madil90) [reply monai/data/writers/niftiwriter.py]: @wyli This is being updated with tests in clara. I'll create a new issue to update this functionality next week.

Comment (madil90) [reply monai/data/writers/niftiwriter.py]: @wyli @ericspod submitted a pull request to this branch that fixes closest_canonical case. Basically we need to store the original affine matrix and then use that to calculate the orientation matrix and reverse values using that. See #74

File: monai/data/readers/niftireader.py (hunk 1 above)
Comment (madil90) [reply monai/data/readers/niftireader.py]: @wyli #74  also uses metadata to save original affine matrix. This is used to reverse the canonical transformation. I'll keep this in mind in case pipeline changes.

**Comment by wyli** [comment]
@ericspod I think the PR is ready, could you please review? after merging this, we could have another iteration to refactor `dense_patch_slices`
